lib/spawn.c

```diff
@@ -69,7 +69,7 @@ int run_command (const char *cmd, const char *argv[],
 	do {
 		wpid = waitpid (pid, status, 0);
 	} while (   ((pid_t)-1 == wpid && errno == EINTR)
-	         || (wpid != pid));
+	         || ((pid_t)-1 != wpid && wpid != pid));
```

Comment:
Since pid cannot be -1 here (we checked for that above), the extra check for wpid != -1 seems redundant.

Reply:
You're right that according to the waitpid manpage the wpid == -1 && errno != EINTR condition cannot happen because the other possible errors are ECHILD and EINVAL which should not happen due to the parameters used and no SIGCHLD set to SIG_IGN. But if there is another undocumented permanent waitpid() error, without this added condition the loop will hang forever as wpid == -1 and pid !=-1 so wpid != pid is true and the loop continues forever.